### PR TITLE
Other formats beyond "label (unit)"

### LINKED ISSUE #33: Square backets
Is there any reference to why normal brackets `()` are used to denote the units on the plot axes? I am used to the convension of square backets `[]`, and the only think I found from quick googling is that one should denote e.g. time as `time / s`, which seems worse to me. Could there potentially be a flag or optional argument one can use to change the default?

### PR DESCRIPTION
**Pull request by gustaphe**
closes #33 

Allows a keyword argument `surround`, which gives near-perfect control of the label formatting. 

```julia
plot(x,y,xlabel="Length",surround=:slash)
```
gives label `"Length / m"`. `surround` can be a number of symbols like this, a tuple of delimeters, or a function that transforms two arguments `(label, unit)` into a string.

### CODE REVIEW

**Comment by gustaphe** [comment]
And the killer app:

```julia
using Unitful,UnitfulRecipes,UnitfulLatexify,Latexify,Plots # Here's to hoping for a future with first class optional dependencies so we don't have to do this...
labelformatter(l,u) = "\$$l / $(latexify(u,env=:raw))\$"
E = rand(10)u"kg*m/s^2"
v = rand(10)u"m/s"
plot(v,E,st=:scatter,xlabel="v",ylabel="E",surround=labelformatter)
```

labelformatter

File: src/UnitfulRecipes.jl
Comment (briochemc) [inline src/UnitfulRecipes.jl]: Interesting, I think I would have done this with a constant `Dict` and then do something like
```julia
surround(l, u, f::Symbol) = surround(l, u, SURROUND_DICT[f])
```
Is this function better than a `Dict` for some reason? (Not asking to change this, just curious)

**Comment by briochemc** [comment]
Nice work, thanks! Do you think you could come up with short examples to add to the docs? Maybe one of them could span all the symbols, like
```julia
x = rand(10) * u"m/s"
symbols = [:round, :square, :curly, :angle, :slash, :slashround, :slashsquare, :slashcurly, :slashangle, :verbose]
plot([plot(x, surround=s) for s in symbols]...)
```
Whatever examples you come up with, could you post a screenshot here? (I don't think we get docs previews from external PRs unfortunately.)

File: docs/lit/examples/1_Examples.jl
Comment (briochemc) [inline docs/lit/examples/1_Examples.jl]: I think the Oxford comma would be helpful here! (Also typo on "customizatbility
".) What about something a bit shorter, like:
```
# For more customizability, a `Char`, a `String`, or a `Tuple`
# (of `Char`s or `String`s) can be supplied. They will be inserted 
# differently around the unit depending
# on the length of the tuple. See these examples:
```
And then we have more examples that show these behaviors with tuples

File: docs/lit/examples/1_Examples.jl
Comment (briochemc) [inline docs/lit/examples/1_Examples.jl]: I think it would be good to cut the paragraph after "parentheses." here and have Julia example blocks with these symbols.

Comment (briochemc) [reply docs/lit/examples/1_Examples.jl]: Maybe good place for something like I suggested in the conversation, with many subplots each corresponding to a different symbol, but maybe with `"surround=symbol"` as the title, like
```julia
symbols = [:round, :square, :curly, :angle, :slash, :slashround, :slashsquare, :slashcurly, :slashangle, :verbose, nothing]
plot([plot(y, title="surround=$s", surround=s) for s in symbols]...)
```

File: docs/lit/examples/1_Examples.jl
Comment (briochemc) [inline docs/lit/examples/1_Examples.jl]: Could you add spaces after commas for consistency with the rest of the formatting?
```julia
plot(y, ylabel="mass", surround=:square)
```

File: docs/lit/examples/1_Examples.jl
Comment (briochemc) [inline docs/lit/examples/1_Examples.jl]: spaces after commas:
```julia
plot(y, ylabel="mass", surround=(l,u)->string("\$\\frac{",l,"}{\\mathrm{",u,"}}\$"))
```

File: src/UnitfulRecipes.jl
Comment (briochemc) [inline src/UnitfulRecipes.jl]: spaces after commas:
```julia
        attr[key] = UnitfulString(S(surround(label, u, get(attr, :surround, :round))), u)
```

File: src/UnitfulRecipes.jl
Comment (briochemc) [inline src/UnitfulRecipes.jl]: This looks neat and exhaustive: nice work! Otherwise, spaces after commas :) :
```julia
surround(l, u, f::Nothing) = string(l, ' ', u)
surround(l, u, f::Function) = f(l, u)
surround(l, u, f::AbstractString) = string(l, f, u)
surround(l, u, f::NTuple{2,<:AbstractString}) = string(l, f[1], u, f[2])
surround(l, u, f::NTuple{3,<:AbstractString}) = string(f[1], l, f[2], u, f[3])
surround(l, u, f::Char) = string(l, ' ', f, ' ', u)
surround(l, u, f::NTuple{2,Char}) = string(l, ' ', f[1], u, f[2])
surround(l, u, f::NTuple{3,Char}) = string(f[1], l, ' ', f[2], u, f[3])
surround(l, u, f::Bool) = f ? surround(l, u, :round) : surround(l, u, nothing)
```

File: test/runtests.jl
Comment (briochemc) [inline test/runtests.jl]: Thanks for the set of tests! Spaces after commas (check that I did not mess up anything!):
```julia
        args = (x, y)
        kwargs = (:xlabel=>"hello", :ylabel=>"hello")
        @test yguide(plot(args...; kwargs..., surround=nothing)) == "hello s"
        @test yguide(plot(args...; kwargs..., surround=(l,u) -> string(u, " is the unit of ", l))) == "s is the unit of hello"
        @test yguide(plot(args...; kwargs..., surround=", dear ")) == "hello, dear s"
        @test yguide(plot(args...; kwargs..., surround=(", dear ", " esq."))) == "hello, dear s esq."
        @test yguide(plot(args...; kwargs..., surround=("well ", ", dear ", " esq."))) == "well hello, dear s esq."
        @test yguide(plot(args...; kwargs..., surround='?')) == "hello ? s"
        @test yguide(plot(args...; kwargs..., surround=('<', '>'))) == "hello <s>"
        @test yguide(plot(args...; kwargs..., surround=('A', 'B', 'C'))) == "Ahello BsC"
        @test yguide(plot(args...; kwargs..., surround=false)) == "hello s"
        @test yguide(plot(args...; kwargs..., surround=true)) == "hello (s)"
        @test yguide(plot(args...; kwargs..., surround=:round)) == "hello (s)"
        @test yguide(plot(args...; kwargs..., surround=:square)) == "hello [s]"
        @test yguide(plot(args...; kwargs..., surround=:curly)) == "hello {s}"
        @test yguide(plot(args...; kwargs..., surround=:angle)) == "hello <s>"
        @test yguide(plot(args...; kwargs..., surround=:slash)) == "hello / s"
        @test yguide(plot(args...; kwargs..., surround=:slashround)) == "hello / (s)"
        @test yguide(plot(args...; kwargs..., surround=:slashsquare)) == "hello / [s]"
        @test yguide(plot(args...; kwargs..., surround=:slashcurly)) == "hello / {s}"
        @test yguide(plot(args...; kwargs..., surround=:slashangle)) == "hello / <s>"
        @test yguide(plot(args...; kwargs..., surround=:verbose)) == "hello in units of s"
```

File: src/UnitfulRecipes.jl
Comment (gustaphe) [reply src/UnitfulRecipes.jl]: I don't know. A dict has a different memory layout, but is probably indexed more efficiently (this is essentially a linear search). But like this we can order things by how likely they are to be used, so every time someone uses `:round` (which is the default) it's just a single comparison.

I don't think there's a measurable difference in performance for this case.

**Comment by gustaphe** [comment]
alllabelformats

```julia
labelformatter(l, u) = string(u, ' ', l)
surrounds = [
             [
              :round,
              :square,
              :curly,
              :angle,
              :slash,
              :slashround,
              :slashsquare,
              :slashcurly,
              :slashangle,
              :verbose,
             ],
             [
              ", in ",
              (", in (", ")"),
              ("[", "] = (", ")"),
              ':',
              ('$', '$'),
              (':', ':', ':'),
             ],
             [
              nothing,
              labelformatter,
              false,
              true,
             ],
            ]
```

Unfortunately, `repr(::Function)` is not very pleasing to the eye. But I guess this is a decent lookup table if someone needs it?

**Comment by gustaphe** [comment]
> Unfortunately, `repr(::Function)` is not very pleasing to the eye.

this is in reference to something that happens when the example is run outside of the REPL. It says `Main.ex-1_Examples.labelformatter` instead of just `formatter`. I tried with an anonymous function but it's way worse (`var"#7#8"()`)

File: src/UnitfulRecipes.jl
Comment (jw3126) [reply src/UnitfulRecipes.jl]: FWIW I would also have used a `Dict` for such a switch that is not a performance bottleneck. Why? Because a `Dict` is easier to inspect (e.g. `display` gives a ton of information) than a function. 
Anyway it is not a huge deal and you can leave it as is.

**Review by jw3126 — COMMENTED** [review]
Awsome PR @gustaphe big thanks! Some bikeshedding, do we want to call this keyword surround? Is this term used elsewhere for something similar? Alternatives I have in mind:
`formatunit`, `formatguide`, `guidestyle`

**Comment by gustaphe** [comment]
It's not bikeshedding if it's important :p

I'm not married to `surround`, that's just what this was when I started out.

I would reverse the order, `unitformat` (noun) is better than `formatunit` (verb?).

I think there's two ways to do it: either we steer clear of anything that sounds like it could be Plots functionality (so we mention "unit" or "quantitylabelformat" or something), or we lean into it and use `labelformat`, being prepared for the eventuality that this is introduced in Plots.

Hell, using `applicable(labelformat, s, u)` we could even allow single-argument labelformats, filling in this lacuna for now.

I think I'm spiralling. `unitlabelformat`?

**Comment by jw3126** [comment]
> I think there's two ways to do it: either we steer clear of anything that sounds like it could be Plots functionality (so we mention "unit" or "quantitylabelformat" or something), or we lean into it and use `labelformat`, being prepared for the eventuality that this is introduced in Plots.

That is a very good point. Yes I think we should try to stay out of `Plots.jl`s way and use a name that contains `unit`. `unitformat` sounds good to me, what do you think @briochemc ?

**Comment by briochemc** [comment]
Just wanted to say I'm going to be too busy to contribute anything in the next few days (starting now), so feel free to go ahead without me, or if you want to wait, I'll be happy to chime in next week 😃

**Comment by gustaphe** [comment]
It was becoming enough extra characters of code that I took you up on the `Dict` suggestion

File: docs/lit/examples/1_Examples.jl
Comment (jw3126) [inline docs/lit/examples/1_Examples.jl]:
```suggestion
# ### Unit formatting
```

**Review by jw3126 — APPROVED** [review]
Looks good thanks again @gustaphe for the great PR!

File: docs/lit/examples/1_Examples.jl
Comment (gustaphe) [reply docs/lit/examples/1_Examples.jl]: Oh man, I just did `vim $(grep -Rl surround)`, I forgot about human-readable formats.

**Comment by briochemc** [comment]
Looks great to me too! I would do 2 tiny last things. 
- I would simplify the example that goes in the docs further, and I think we can use the `UNIT_FORMATS::Dict` you added for that (I'll put a comment on the relevant lines of code). 
- I would only apply the style changes of adding a space after comma on those lines of code that pertain to this PR so that this PR is about the cool new feature only (I think that's just a few reversions). Note I don't feel too strongly about this but I think eventually what we want to do is adopt one of the "established" formatting styles out there (e.g., [BlueStyle](https://github.com/invenia/BlueStyle) or another one) and stick to it, and that could be put in an issue with "good first PR" or something?

File: docs/lit/examples/1_Examples.jl
Comment (briochemc) [inline docs/lit/examples/1_Examples.jl]: I think we can turn this into something cleaner with much shorter code cells, e.g., something like this (not tested, likely contains typos/errors):

```julia
# ### Unit formatting

# If you prefer some other formatting over the round parentheses, you can
# supply a keyword `unitformat`, which can be a number of different things.

# `unitformat` can be a boolean or `nothing`:

plot([plot(y, ylab="mass", title=repr(s), unitformat=s) for s in (nothing, true, false)]...)

# `unitformat` can be a symbol from one of the predefined symbols that you can check with

URsymbols = keys(UnitfulRecipes.UNIT_FORMATS)

# and allow you one of these unit formats:

plot([plot(y, ylab="mass", title=repr(s), unitformat=s) for s in URsymbols]...)

# `unitformat` can also be a `Char`, a `String`, or a `Tuple` 
# (of `Char`s or `String`s), which will be inserted
# arround the label and unit depending on the length of the tuple

URtuples = [", in ", (", in (", ")"), ("[", "] = (", ")"), ':', ('$', '$'), (':', ':', ':')],
plot([plot(y, ylab="mass", title=repr(s), unitformat=s) for s in URtuples]...)

# And for *extreme* customizability, `unitformat` can a function that takes two
# arguments, `(label, unit)`, and returns a string:

plot(y, ylab="mass", unitformat=(l, u)->string("\$\\frac{", l, "}{\\mathrm{", u, "}}\$"))
```

**Review by briochemc — COMMENTED** [review]
I added a suggestion for the example to go in the docs that I think will make it cleaner. Please don't just copy-paste it as I have ***not*** tested it and it likely has some typos and errors!

**Comment by briochemc** [comment]
> > I think there's two ways to do it: either we steer clear of anything that sounds like it could be Plots functionality (so we mention "unit" or "quantitylabelformat" or something), or we lean into it and use `labelformat`, being prepared for the eventuality that this is introduced in Plots.
> 
> That is a very good point. Yes I think we should try to stay out of `Plots.jl`s way and use a name that contains `unit`. `unitformat` sounds good to me, what do you think @briochemc ?

I agree that if this integrates somewhere it would integrate with Unitful.jl rather than Plots.jl. And I like `unitformat` best anyway! 😄

File: docs/lit/examples/1_Examples.jl
Comment (gustaphe) [reply docs/lit/examples/1_Examples.jl]: That looks a lot better. I just changed a couple of formulations that weren't clear to me.

**Comment by gustaphe** [comment]
> > > I think there's two ways to do it: either we steer clear of anything that sounds like it could be Plots functionality (so we mention "unit" or "quantitylabelformat" or something), or we lean into it and use `labelformat`, being prepared for the eventuality that this is introduced in Plots.
> > 
> > 
> > That is a very good point. Yes I think we should try to stay out of `Plots.jl`s way and use a name that contains `unit`. `unitformat` sounds good to me, what do you think @briochemc ?
> 
> I agree that if this integrates somewhere it would integrate with Unitful.jl rather than Plots.jl. And I like `unitformat` best anyway!

I could easily see Plots.jl introducing a `guideformatter` option, and then hijacking that to work with two-argument functions would be pretty neat. But yes, let's just stay out of the way.

**Comment by gustaphe** [comment]
> I think eventually what we want to do is adopt one of the "established" formatting styles out there (e.g., [BlueStyle](https://github.com/invenia/BlueStyle) or another one) and stick to it, and that could be put in an issue with "good first PR" or something?

I read up on it, and one could achieve most of this by running
```julia
using JuliaFormatter
format(".", BlueStyle())
```
in the package root. I'd make a PR out of it, but in case there are any personal preferences maybe it's the kind of thing best left to maintainers themselves.

**Comment by jw3126** [comment]
@gustaphe I can do the JuliaFormatter PR after this one, to avoid merge conflicts.

**Comment by jw3126** [comment]
Great PR @gustaphe ! From my pov, this is ready to be merged. Any objections anybody?

**Review by briochemc — APPROVED** [review]

**Comment by briochemc** [comment]
> Great PR @gustaphe ! From my pov, this is ready to be merged. Any objections anybody?

No objections your honor! 😉

**Comment by briochemc** [comment]
Oh actually just one last thing, @gustaphe can you bump the version number so we can tag a release directly on this?

**Comment by KronosTheLate** [comment]
As the author of issue #33, I would like to extend a huge thanks for alle the good work put into resolving it. You guys are great :D

**Pull request merged**
